File: src/lib/api/usageStats.ts
import {
  APP_SERVER_METHOD_USAGE_STATS_DAILY_TRENDS_LIST,
  APP_SERVER_METHOD_USAGE_STATS_MODEL_RANKING_LIST,
  APP_SERVER_METHOD_USAGE_STATS_READ,
  createAppServerClient,
} from "./appServer";

export interface UsageStatsResponse {
  total_conversations: number;
  total_messages: number;
  total_tokens: number;
  total_time_minutes: number;
  monthly_conversations: number;
  monthly_messages: number;
  monthly_tokens: number;
  today_conversations: number;
  today_messages: number;
  today_tokens: number;
}

export interface ModelUsage {
  model: string;
  conversations: number;
  tokens: number;
  percentage: number;
}

export interface DailyUsage {
  date: string;
  conversations: number;
  tokens: number;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === "object" && !Array.isArray(value);
}

function isNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function isString(value: unknown): value is string {
  return typeof value === "string" && value.trim().length > 0;
}

export async function getUsageStats(
  timeRange: string,
): Promise<UsageStatsResponse> {
  const result = (await createAppServerClient().readUsageStats({ timeRange }))
    .result.stats;
  if (
    !isRecord(result) ||
    !isNumber(result.totalConversations) ||
    !isNumber(result.totalMessages) ||
    !isNumber(result.totalTokens) ||
    !isNumber(result.totalTimeMinutes) ||
    !isNumber(result.monthlyConversations) ||
    !isNumber(result.monthlyMessages) ||
    !isNumber(result.monthlyTokens) ||
    !isNumber(result.todayConversations) ||
    !isNumber(result.todayMessages) ||
    !isNumber(result.todayTokens)
  ) {
    throw new Error(
      `${APP_SERVER_METHOD_USAGE_STATS_READ} 未返回有效使用统计数据`,
    );
  }
  return {
    total_conversations: result.totalConversations,
    total_messages: result.totalMessages,
    total_tokens: result.totalTokens,
    total_time_minutes: result.totalTimeMinutes,
    monthly_conversations: result.monthlyConversations,
    monthly_messages: result.monthlyMessages,
    monthly_tokens: result.monthlyTokens,
    today_conversations: result.todayConversations,
    today_messages: result.todayMessages,
    today_tokens: result.todayTokens,
  };
}

export async function getModelUsageRanking(
  timeRange: string,
): Promise<ModelUsage[]> {
  const result = (
    await createAppServerClient().listUsageStatsModelRanking({ timeRange })
  ).result.ranking;
  if (
    !Array.isArray(result) ||
    result.some(
      (item) =>
        !isRecord(item) ||
        !isString(item.model) ||
        !isNumber(item.conversations) ||
        !isNumber(item.tokens) ||
        !isNumber(item.percentage),
    )
  ) {
    throw new Error(
      `${APP_SERVER_METHOD_USAGE_STATS_MODEL_RANKING_LIST} 未返回有效模型使用排行`,
    );
  }
  return result as ModelUsage[];
}

export async function getDailyUsageTrends(
  timeRange: string,
): Promise<DailyUsage[]> {
  const result = (
    await createAppServerClient().listUsageStatsDailyTrends({ timeRange })
  ).result.trends;
  if (
    !Array.isArray(result) ||
    result.some(
      (item) =>
        !isRecord(item) ||
        !isString(item.date) ||
        !isNumber(item.conversations) ||
        !isNumber(item.tokens),
    )
  ) {
    throw new Error(
      `${APP_SERVER_METHOD_USAGE_STATS_DAILY_TRENDS_LIST} 未返回有效每日使用趋势`,
    );
  }
  return result as DailyUsage[];
}

File: lime-rs/crates/app-server/src/local_data_source/usage_stats.rs
use app_server_protocol::UsageStatsDailyTrendsListResponse;
use app_server_protocol::UsageStatsDailyUsage;
use app_server_protocol::UsageStatsModelRankingListResponse;
use app_server_protocol::UsageStatsModelUsage;
use app_server_protocol::UsageStatsRangeParams;
use app_server_protocol::UsageStatsReadResponse;
use app_server_protocol::UsageStatsSummary;
use lime_core::database;
use lime_core::database::DbConnection;
use lime_services::usage_statistics_service;

pub(crate) fn read_usage_stats(
    db: &DbConnection,
    params: UsageStatsRangeParams,
) -> Result<UsageStatsReadResponse, String> {
    let conn = database::lock_db(db).map_err(|error| error.to_string())?;
    let stats = usage_statistics_service::get_usage_stats_from_db(&params.time_range, &conn)
        .map_err(|error| error.to_string())?;
    Ok(UsageStatsReadResponse {
        stats: UsageStatsSummary {
            total_conversations: stats.total_conversations,
            total_messages: stats.total_messages,
            total_tokens: stats.total_tokens,
            total_time_minutes: stats.total_time_minutes,
            monthly_conversations: stats.monthly_conversations,
            monthly_messages: stats.monthly_messages,
            monthly_tokens: stats.monthly_tokens,
            today_conversations: stats.today_conversations,
            today_messages: stats.today_messages,
            today_tokens: stats.today_tokens,
        },
    })
}

pub(crate) fn list_usage_stats_model_ranking(
    db: &DbConnection,
    params: UsageStatsRangeParams,
) -> Result<UsageStatsModelRankingListResponse, String> {
    let conn = database::lock_db(db).map_err(|error| error.to_string())?;
    let ranking =
        usage_statistics_service::get_model_usage_ranking_from_db(&params.time_range, &conn)
            .map_err(|error| error.to_string())?
            .into_iter()
            .map(|item| UsageStatsModelUsage {
                model: item.model,
                conversations: item.conversations,
                tokens: item.tokens,
                percentage: item.percentage,
            })
            .collect();
    Ok(UsageStatsModelRankingListResponse { ranking })
}

pub(crate) fn list_usage_stats_daily_trends(
    db: &DbConnection,
    params: UsageStatsRangeParams,
) -> Result<UsageStatsDailyTrendsListResponse, String> {
    let conn = database::lock_db(db).map_err(|error| error.to_string())?;
    let trends =
        usage_statistics_service::get_daily_usage_trends_from_db(&params.time_range, &conn)
            .map_err(|error| error.to_string())?
            .into_iter()
            .map(|item| UsageStatsDailyUsage {
                date: item.date,
                conversations: item.conversations,
                tokens: item.tokens,
            })
            .collect();
    Ok(UsageStatsDailyTrendsListResponse { trends })
}
